### ISSUE TITLE
Support local Lambda invocations using the FunctionName

### ISSUE THREAD

**Issue by kylevoyto** [awslabs/aws-sam-cli#1025]
### Proposal

Support local Lambda invocations based on the FunctionName in addition to the Logical ID.

### Issue

The following Lambda function is defined in my SAM template:

```
Resources:
  MainFunction:
    Type: AWS::Serverless::Function
    Properties:
      FunctionName: org-project-main
      CodeUri: functions/main/
      Handler: main.lambda_handler
      Timeout: 15
```

I want to invoke the Lambda function locally using the name defined in the FunctionName property:

```
sam local invoke "org-project-main" --no-event
```

But the error is produced:

```
2019-02-25 13:25:19 org-project-main not found. Possible options in your template: ['MainFunction', 'OtherFunction']
Error: Function org-project-main not found in template
```

We follow the convention at my institution to name Lambda functions like org-project-name, which is allowed by Lambda, but the Logical ID of the resource (MainFunction in this case) must be alphanumeric. Therefore, I cannot set the Logical ID and the FunctionName to the same value.

### Additional Details

I am using `sam --version`:

```
SAM CLI, version 0.11.0
```

On macOS `sw_vers`:

```
ProductName:    Mac OS X
ProductVersion: 10.14.3
BuildVersion:   18D109
```

**Comment by matrinox** [comment]
I second this. We use dash case for lambda function names as well. Using `FunctionName` works when deployed but not when run locally. This causes some inconsistencies between development & AWS envs as application code run on AWS doesn't match the name run in development. In development, we start the lambda function using `sam local start-lambda`.

**Comment by matrinox** [comment]
@kylevoyto I got it to work by changing the resource type to `AWS::Lambda::Function` and that allowed me to use dash case for the logical ID. Problem solved :). Oh, and you have to change `CodeUri` to `Code` and probably a few other differences that didn't impact me.

Still, I think this might be a bug or at least an inconsistency. You would expect the lambda name in AWS to be the same locally.

**Comment by kylevoyto** [comment]
@matrinox thanks for sharing the work-around!

**Comment by matrinox** [comment]
@kylevoyto Sorry for misleading. I tried to deploy the code and I saw the "Resource name ... is non alphanumeric" error :(. Sam CLI needs some fixing

**Comment by jfuss** [comment]
Released in [v0.46.0](https://github.com/awslabs/aws-sam-cli/releases/tag/v0.46.0).

Closing
